### PR TITLE
Deflake Tracing e2e test

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Please include a summary of the changes.

This bug fixes ... \ This new feature can be used to ...

# Context

Users ran into this bug doing ... \ Users needed this feature to ...

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [ ] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sam-heilbron** [comment]
/skip-changelog there is no current github issue tracking this flake

**Review by nfuden — APPROVED** [review]
This seems reasonable to me. I am still not sure from the example logs that it will fully be fixed by this but it seems to at least make it much less flakey.

File: test/e2e/zipkin_test.go [hunk 1]
```diff
@@ -114,304 +108,278 @@ var _ = Describe("Zipkin config loading", func() {
...
-
-			// write gateways and wait for them to be created
-			err := gloohelpers.WriteDefaultGateways(writeNamespace, testClients.GatewayClient)
-			Expect(err).NotTo(HaveOccurred(), "Should be able to write default gateways")
-			Eventually(func() (gatewayv1.GatewayList, error) {
-				return testClients.GatewayClient.List(writeNamespace, clients.ListOpts{})
-			}, "10s", "0.1s").Should(HaveLen(2), "Gateways should be present")
-
-			// run envoy
-			err = envoyInstance.RunWithConfigFile(testClients.GlooPort, "./envoyconfigs/zipkin-static-cluster.yaml")
```
Comment (gunnar-solo) [inline test/e2e/zipkin_test.go:340]: what was the rationale behind moving from a config yaml to in-momory declaration?

File: test/e2e/gzip_test.go [hunk 2]
```diff
@@ -55,152 +57,118 @@ var _ = Describe("gzip", func() {
...
-		}, "5s", "0.1s").ShouldNot(BeNil())
-	}
-
-	checkVirtualService := func(testVs *gatewayv1.VirtualService) {
-		Eventually(func() (*gatewayv1.VirtualService, error) {
-			return testClients.VirtualServiceClient.Read(testVs.Metadata.GetNamespace(), testVs.Metadata.GetName(), clients.ReadOpts{})
-		}, "5s", "0.1s").ShouldNot(BeNil())
-	}
-
-	testRequest := func(jsonStr string) string {
```
Comment (gunnar-solo) [inline test/e2e/gzip_test.go:97]: it seems to me that the bulk of the fix comes from wrapping this in an `Eventually`.  I see that we're also pulling in a bunch of "previously discussed best practices" as well

(just an observation)

File: test/e2e/gzip_test.go [hunk 3]
```diff
@@ -55,152 +57,118 @@ var _ = Describe("gzip", func() {
...
 
-		// write a test upstream
 		// this is the upstream that will handle requests
-		testUs := v1helpers.NewTestHttpUpstream(ctx, envoyInstance.LocalAddr())
-		up = testUs.Upstream
-		_, err = testClients.UpstreamClient.Write(up, clients.WriteOpts{OverwriteExisting: true})
-		Expect(err).NotTo(HaveOccurred())
+		testUpstream = v1helpers.NewTestHttpUpstream(ctx, envoyInstance.LocalAddr())
+
+		vsToTestUpstream := gloohelpers.NewVirtualServiceBuilder().
```
Comment (gunnar-solo) [inline test/e2e/gzip_test.go:70]: I appreciate this helper method builder its miles

**Review by bewebi — APPROVED** [review]
Looks good

Github really made a mess of the diff, so it was hard to follow
Part of this is seemingly due to some refactoring/cleanup that wasn't explicitly called out in the PR description
It would be good to mention _all_ changes made to code (ie including refactoring, renaming, etc.) in the description in order to make it easier for reviewers to parse the diff

File: test/e2e/zipkin_test.go (hunk 1 above)
Comment (sam-heilbron) [reply test/e2e/zipkin_test.go:340]: There we 2 tests that did the same thing so I just removed one. One test wrote a cluster to the bootstrap config of envoy (to validate that we can reference the cluster by name), the other write an upstream and then referenced that upstream by cluster name.

I prefer the second type of test, since it validate the gloo pipeline (translation) and is more similar to what a customer would use

File: test/e2e/gzip_test.go (hunk 2 above)
Comment (sam-heilbron) [reply test/e2e/gzip_test.go:97]: Yes that's fair. The eventually is the retry logic that we may need, but also the way that configuration is applied caused the flake. Before we would write a bunch of resources and then modify one. Since we previously had valid resources (gw, vs, upstream) requests would still go through using the old config. The new format is to just write resources once, and only with the config we need.

So yeah, theoretically we could have just used the eventually, but I figured moving more of our tests towards the snapshot model would be easier to reason through

File: test/e2e/gzip_test.go (hunk 3 above)
Comment (sam-heilbron) [reply test/e2e/gzip_test.go:70]: ha me too! I foud that we have so many combinations of VS config, that we tended to have a telescoping constructor anti-pattern:

https://knowledgebasement.com/avoid-the-telescoping-anti-pattern-in-java/

**Comment by sam-heilbron** [comment]
> Looks good
> 
> Github really made a mess of the diff, so it was hard to follow Part of this is seemingly due to some refactoring/cleanup that wasn't explicitly called out in the PR description It would be good to mention _all_ changes made to code (ie including refactoring, renaming, etc.) in the description in order to make it easier for reviewers to parse the diff

Yeah that's fair! The refactoring isn't new per se, in that its a pattern used in other tests that I was trying to bring to these. But good callout that I can make that more explicit in the PR context
